apps/news-site/news-site-next/src/context/data-context.js

```diff
@@ -7,20 +7,74 @@ const DEFAULT_DIR = "ltr";
 
 const DataContext = createContext(null);
 
+let ignore = false;
+
+async function getAndInitializeStaticConfig(url) {
+    const response = await fetch(url);
+    const data = await response.json();
+    return data;
+}
+
 export const DataContextProvider = ({ children }) => {
-    const urlParams = new URLSearchParams(window.location.search);
-    const langFromUrl = urlParams.get("lang")?.toLowerCase();
-    const lang = langFromUrl && langFromUrl in dataSource ? langFromUrl : DEFAULT_LANG;
-    const dir = lang && RTL_LOCALES.includes(lang) ? "rtl" : DEFAULT_DIR;
-
-    document.documentElement.setAttribute("dir", dir);
-    document.documentElement.setAttribute("lang", lang);
-
-    const value = {
-        lang,
-        dir,
-        ...dataSource[lang],
-    };
+    const [value, setValue] = useState({});
+    // initialized should persist between re-renders
+    const initialized = useRef(false);
+
+    // ignore always resets on re-render, hence global var
+    ignore = false;
+
+    useEffect(() => {
+        if (initialized.current)
+            return;
+
+        /**
+         * read lang & dir
+         */
+        const urlParams = new URLSearchParams(window.location.search);
+        const langFromUrl = urlParams.get("lang")?.toLowerCase();
+        const lang = langFromUrl && langFromUrl in dataSource ? langFromUrl : DEFAULT_LANG;
+        const dir = lang && RTL_LOCALES.includes(lang) ? "rtl" : DEFAULT_DIR;
+
+        document.documentElement.setAttribute("dir", dir);
+        document.documentElement.setAttribute("lang", lang);
+
+        /**
+         * read config and fetch
+         */
+
+        const configParam = urlParams.get("config");
+
+        /**
+         * decide to load config
+         */
+        if (configParam) {
+            getAndInitializeStaticConfig(configParam)
+                .then(data => {
+                    if (ignore)
+                        return;
+
+                    setValue({
+                        lang,
+                        dir,
+                        ...dataSource[lang],
+                        config: { ...data },
+                    });
+                });
+        } else {
+            setValue({
+                lang,
+                dir,
+                ...dataSource[lang],
+            });
+        }
+
+        initialized.current = true;
+
+        // eslint-disable-next-line consistent-return
+        return () => {
+            ignore = true;
+        };
+    }, []);
```

Comment:
I would propose something like this:

```javascript
let initConfigPromise = null;
function getAndInitializePageConfig() {
    initConfigPromise ??= (async () => {
        /**
         * read lang & dir
         */
        const urlParams = new URLSearchParams(window.location.search);
        const langFromUrl = urlParams.get("lang")?.toLowerCase();
        const lang = langFromUrl && langFromUrl in dataSource ? langFromUrl : DEFAULT_LANG;
        const dir = lang && RTL_LOCALES.includes(lang) ? "rtl" : DEFAULT_DIR;

        document.documentElement.setAttribute("dir", dir);
        document.documentElement.setAttribute("lang", lang);

        /**
         * read config and fetch
         */

        const configParam = urlParams.get("config");
        let config = {};

        /**
         * decide to load config
         */
        if (configParam) {
            const response = await fetch(configParam);
            config = await response.json();
        }

        return {
            lang,
            dir,
            ...dataSource[lang],
            ...config,
        };
    })();
    return initConfigPromise;
}

export const DataContextProvider = ({ children }) => {
    const [value, setValue] = useState({});

    useEffect(() => {
        let ignore = false;
        getAndInitializePageConfig().then(config => {
            if (!ignore) {
                setValue(config);
            }
        });

        return () => {
            ignore = true;
        };
    }, []);

    return <DataContext.Provider value={value}>{children}</DataContext.Provider>;
};
```

IMO, this just makes it more plainly obvious that the `useEffect` is written properly.